### PR TITLE
Support the RTDB emulator

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This adds support for the emulator and also for the "ns" query param.\

Technically, iOS doesn't need the "ns" query param since "localhost" is a valid RTDB namespace name and we could just use "localhost" as the namespace. But since I just went all out on Android (https://github.com/firebase/firebase-android-sdk/pull/680), I decided to make this PR equally awesome. Plus who doesn't like some good URL decoding/encoding mess.

### CODE REVIEW

**Comment by yuchenshi** [comment]
We still need people to be able to specify the database name (for example to test triggers), so just accepting `localhost` is not a great answer. So, thanks for the PR and it's really important!

File: Firebase/Database/Utilities/FUtilities.m [hunk 1]
```diff
@@ -122,54 +122,81 @@ + (NSString *) decodePath:(NSString *)pathString {
...
-    else {
-        [NSException raise:@"No Firebase database specified." format:@"No Firebase database found for input: %@", url];
+        if (!namespace) {
+            namespace = [parts[0] lowercaseString];
+        }
     }
 
+
+    NSString* pathString = [self decodePath:[NSString stringWithFormat:@"/%@", originalPathString]];
+    FPath *  path = [[FPath alloc] initWith:pathString];
```
Comment (mikelehen) [inline Firebase/Database/Utilities/FUtilities.m]: seems like there are extra spaces here?  Why does clang-format not fix that?

File: Firebase/Database/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,10 @@
+# Unreleased
+- [feature] The SDK adds support for the Firebase Database Emulator. To connect
+  to the emulator, specify "http://<emulator_host>/" as your Database URL
+  (via `Database.database(url:)`).
+  If you refer to your emulator host by IP rather than by domain name, you may
+  also need to specify a namespace ("http://<emulator_host>/?ns=<project_id>").
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply Firebase/Database/CHANGELOG.md]: Done.

File: Firebase/Database/Utilities/FUtilities.m [hunk 3]
```diff
@@ -122,54 +122,81 @@ + (NSString *) decodePath:(NSString *)pathString {
...
+        [NSException raise:@"Failed to parse database URL" format:@"Failed to parse database URL: %@", url];
     }
 
-    NSArray *parts = [host componentsSeparatedByString:@"."];
+    NSString* host = [urlComponents.host lowercaseString];
+    NSString* namespace;
+    bool secure;
+
+    if (urlComponents.port != nil) {
+        secure = [urlComponents.scheme isEqualToString:@"https"];
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply Firebase/Database/Utilities/FUtilities.m]: I added support for "wss".

File: Firebase/Database/Utilities/FUtilities.m [hunk 4]
```diff
@@ -122,54 +122,81 @@ + (NSString *) decodePath:(NSString *)pathString {
...
+        host = [host stringByAppendingFormat:@":%@", urlComponents.port];
+    } else if ( [urlComponents.host isEqualToString:@"localhost"]) {
+        secure = NO;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply Firebase/Database/Utilities/FUtilities.m]: Seems like this is not winning anyone over, so I dropped this from the iOS port (will drop the special casing for Android too).

File: Firebase/Database/Utilities/FUtilities.m (hunk 1 above)
Comment (schmidt-sebastian) [reply Firebase/Database/Utilities/FUtilities.m]: Fixed manually. Seems like we are excluded from the formatter (but it still takes 10 seconds to run). I'll get the formatting enabled for RTDB in a follow up.

Comment (schmidt-sebastian) [reply Firebase/Database/Utilities/FUtilities.m]: This source is now clang-formatted.

**Pull request merged**